### PR TITLE
refactors and adds JLong, JDecimal and JSet support to JObjectParser in mongo

### PR DESCRIPTION
**Pull request by phiSgr**
Old code wildcard matches `JLong` and produces empty string. This version matches all cases and avoids the wildcard pattern.

This version also changes the parsing of `$oid`, `$regex` and `$uuid` objects, ignoring the `JObject` if there is an error. Please confirm if this behaviour is desired.

### CODE REVIEW

File: mongo/src/main/scala/org/json4s/mongo/JObjectParser.scala [hunk 1]
```diff
@@ -79,57 +80,45 @@ object JObjectParser  {
...
+      case JObject(JField("$dt", JString(s)) :: Nil) =>
+        formats.dateFormat.parse(s)
+      case JObject(JField("$uuid", JString(s)) :: Nil) =>
+        Try { UUID.fromString(s) } toOption
+      case JArray(arr) => Some(parseArray(arr, formats))
+      case JSet(set) =>  Some(parseArray(set.toList, formats))
+      case JObject(jo) => Some(parseObject(jo, formats))
+      case JBool(b) => Some(java.lang.Boolean.valueOf(b))
+      case JInt(n) => Some(renderInteger(n))
+      case JLong(n) => Some(new java.lang.Long(n))
```
Comment (xuwei-k) [inline mongo/src/main/scala/org/json4s/mongo/JObjectParser.scala]: `java.lang.Long.valueOf(n)`

File: mongo/src/main/scala/org/json4s/mongo/JObjectParser.scala [hunk 2]
```diff
@@ -79,57 +80,45 @@ object JObjectParser  {
...
+      case JDouble(n) => Some(new java.lang.Double(n))
```
Comment (xuwei-k) [inline mongo/src/main/scala/org/json4s/mongo/JObjectParser.scala]: `java.lang.Double.valueOf(n)`

**Comment by xuwei-k** [comment]
Could you write some tests?

**Comment by phiSgr** [comment]
@xuwei-k I have added two tests.

**Comment by mingchuno** [comment]
our company has some code that requires the correct behaviour handling JLong, is there anything stopping the change from being merged?

**Comment by mccartney** [comment]
This PR has been a few years old. The idea seems sound, but I guess it's unlikely someone will take care of that. Closing. Thanks to everyone involved.

**Pull request closed without merging**